### DISCUSSION TITLE
How to resolve include / render tag in template that is from database?

### DISCUSSION THREAD

**Discussion by Huai-Kang** [harttle/liquidjs#414] [Q&A] [answered]
Reference to [Liquidjs: Abstract File System](https://liquidjs.com/tutorials/render-file.html#Abstract-File-System), We can load template from database or an object and so on, and how to resolve include / render tag inside those template?

Let's see there is an object like:
```js
const TEMPLATE = {
  "a.com": {
    "home.liquid": "{% include 'product' %}",
    "product.liquid": "xxx",
  },
  "b.com": {
    "home.liquid": "{% include 'product' %}",
    "product.liquid": "xxx",
  },
}
```

And I custom the renderFileSync function:
```js
{
  readFileSync (file: string) {
    const index = file.indexOf('@');
    const [site, fileName] = [file.slice(0, index), file.slice(index + 1)];
    const siteTemplateString = global.__TEMPLATE.get(site);
    if (siteTemplateString !== undefined) {
      const siteTemplate = JSON.parse(siteTemplateString);
      const template = siteTemplate[fileName];
      if (template !== undefined) {
        return template;
      } else throw new Error(`Cannot find ${ fileName } template under site ${ site }.`);
    } else throw new Error(`Site ${ site } is un-exist.`);
  },
  async readFile (file) {
    throw new Error('Please use Liquid.readFileSync instead.');
  },
  existsSync () { return true },
  async exists () { return true },
  resolve(root, file, ext) { return file },
}
```

So I can resolve home page of a.com like this:
```js
router.get('/', async (ctx, next) => {
  const { host } = ctx.request; // such as 'a.com'
  const data = liquidEngine.renderFileSync(`${ host }@home.liquid`);
  ctx.body = data;
  next();
});
```

In custom renderFileSync function, I can correctly get home.liquid of a.com, but there will missing product.liquid, because I lose host info, so how can I store the host info for every render?

**Comment by harttle** [comment, marked as answer]
For the partial reference part, try the recently added [relativeReference](https://liquidjs.com/tutorials/options.html#Partials-Layouts) option (enabled by default). Add both hosts into `root` option, then include partials by relative path, like `{% include './product' %}` instead of `{% include 'product' %}`.

For the cross origin `{% include %}`, maybe you'll need to specify the absolute URL to make it work, like `{% include "https://example.com/product.liquid" %}`. Or, alternatively, you can invent an "abstracted FS" (implement multi domain by pseudo-directory), like:

```
foo.com/foo.liquid -> https://foo.com/<any prefix>/foo
bar.com/bar.liquid -> https://foo.com/<any prefix>/bar
```

Then in `foo.liquid`, include `bar.liquid` by `{% include "../bar.com/bar.liquid" %}`
